pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/errorprone/UnnecessaryCaseChangeRule.java

```diff
@@ -4,83 +4,111 @@
 
 package net.sourceforge.pmd.lang.java.rule.errorprone;
 
-import net.sourceforge.pmd.lang.ast.Node;
+import static java.util.Arrays.asList;
+
+import java.util.List;
+
+import net.sourceforge.pmd.lang.java.ast.ASTArgumentList;
+import net.sourceforge.pmd.lang.java.ast.ASTArguments;
 import net.sourceforge.pmd.lang.java.ast.ASTName;
 import net.sourceforge.pmd.lang.java.ast.ASTPrimaryExpression;
 import net.sourceforge.pmd.lang.java.ast.ASTPrimaryPrefix;
 import net.sourceforge.pmd.lang.java.ast.ASTPrimarySuffix;
+import net.sourceforge.pmd.lang.java.ast.JavaNode;
 import net.sourceforge.pmd.lang.java.rule.AbstractJavaRule;
 
 public class UnnecessaryCaseChangeRule extends AbstractJavaRule {
 
+    private static final List<String> CASE_CHANGING_METHODS = asList("toLowerCase", "toUpperCase");
+    private static final List<String> EQUALITY_METHODS = asList("equals", "equalsIgnoreCase");
+
     @Override
-    public Object visit(ASTPrimaryExpression exp, Object data) {
-        int n = exp.getNumChildren();
-        if (n < 4) {
-            return data;
+    public Object visit(ASTPrimaryExpression expr, Object data) {
+        int caseChangingCallIndex = getCaseChangingMethodCallIndex(expr);
+        if (caseChangingCallIndex != -1) {
+            int chainedMethodCallIndex = caseChangingCallIndex + 2;
+            if (hasEqualsMethodCallChainedAtPosition(expr, chainedMethodCallIndex)
+                    || isArgumentOfEqualsMethodCall(expr)) {
+                addViolation(data, expr);
+            }
         }
+        return super.visit(expr, data);
+    }
 
-        int first = getBadPrefixOrNull(exp, n);
-        if (first == -1) {
-            return data;
+    private int getCaseChangingMethodCallIndex(ASTPrimaryExpression expr) {
+        List<JavaNode> exprNodes = expr.findChildrenOfType(JavaNode.class);
+        for (int callArgsIndex = 1; callArgsIndex < exprNodes.size(); callArgsIndex++) {
+            int callIndex = callArgsIndex - 1;
+            if (isCaseChangingMethodCall(exprNodes.get(callIndex), exprNodes.get(callArgsIndex))) {
+                return callIndex;
+            }
         }
+        return -1;
+    }
 
-        String second = getBadSuffixOrNull(exp, first + 2);
-        if (second == null) {
-            return data;
-        }
+    private boolean isCaseChangingMethodCall(JavaNode methodCall, JavaNode methodCallArgs) {
+        String methodName = getCalledMethodName(methodCall);
+        int methodArgsCount = getCalledMethodArgsCount(methodCallArgs);
+        return methodName != null
+                && isNameOfCaseChangingMethod(methodName) && methodArgsCount == 0;
```

Comment:
I think, it would make sense to integrate the null check into the method `isNameOfCaseChangingMethod`, wdyt?